### ISSUE TITLE
Discussion: Which folder hierarchy should we recommend in project structure?

### ISSUE THREAD

**Issue by joelostblom** [merely-useful/merely-useful.github.io#347]
I would suggest something like the below. It is rather simple and compatible with and can be extended to both the [cookicutter datascience](https://drivendata.github.io/cookiecutter-data-science/#directory-structure) and the [pyscaffold](https://github.com/pyscaffold/pyscaffoldext-dsproject) project structure. Please compare with these and see if something important is missing

```
pkg_name
├── data/
│   ├── processed/          <- Intermediate data that has been transformed.
│   └── raw/                <- The original, immutable data dump.
├── doc/                    <- Directory for documentation.
├── notebooks/              <- Jupyter notebooks. Naming convention is a number (for
│                              ordering) and a description, e.g. `1-initial-data-exploration`.
├── reports/                <- Generated analysis as HTML, PDF, LaTeX, etc.
│   └── fig/                <- Generated plots and figures for reports.
├── pkg_name/               <- Actual Python package where the main functionality goes.
│                              could also be src/pke_name or py/pkg_name.
├─  tests/                  <- Unit tests which can be run with `py.test`.
├── CHANGELOG.md            <- Changelog to keep track of new features and fixes.
├── LICENSE.txt             <- License as chosen on the command-line.
├── README.md               <- The top-level README for developers.
├── environment.yaml        <- The environment file for reproducibility (could be requirements.txt).
└── setup.py                <- Use `python setup.py develop` to install for development or
                               or create a distribution with `python setup.py bdist_wheel`.
```

For R, I imagine that it would look similar to this, @lwjohnst86 @cwickham you discussed project structure a bit in [this previous PR](https://github.com/merely-useful/merely-useful.github.io/pull/203), could you (or anyone else) advice? Ideally we would be able to suggested an almost language independent dir structure, but if the conventions between R and Python are not widely different, this might not make sense, lets discuss.

```
pkg_name
├── data/
│   ├── processed/          <- Intermediate data that has been transformed.
│   └── raw/                <- The original, immutable data dump.
├── doc/                    <- Directory for documentation.
├── notebooks/              <- Rmarkdown notebooks. Naming convention is a number (for
│                              ordering) and a description, e.g. `1-initial-data-exploration`.
├── reports/                <- Generated analysis as HTML, PDF, LaTeX, etc.
│   └── fig/                <- Generated plots and figures for reports.
├── R/                      <- R package and scripts? subdirs?
├── tests/                  <- Unit tests which can be run with `py.test`.
├── CHANGELOG.md            <- Changelog to keep track of new features and fixes.
├── LICENSE.txt             <- License as chosen on the command-line.
├── README.md               <- The top-level README for developers.
└── .Rproj                  <- The environment file for reproducibility.
```

In addition to general discussion, here are a few specific points to consider:

1. Should we have a readme in each dir or just the top dir (which could include a desc of subdirs if needed)?
2. What's the name of the dir where the code goes? Having `src/PKG_NAME` has the advantage of one-off scripts (if there are any) being placed in `src/` (or `src/scripts`) whereas they otherwise need a separate dir (e.g. `scripts/`).
    - For Python, I am strongly in favor of including the package name somewhere and not just calling the dir `src/`. Putting `__init__.py` within this folder means that when you install this as an editable package in your environment, your functions can now be imported as `my_project.module.function` vs `src.module.function`. The latter would cause name collisions if functions are shared between projects, while the former makes it clear where the functions are from.  
3. I am generally in favor of treating things as a package early on, and not messing around too much in another dir structure and then restructuring to make a package. At least in Python, there is little extra effort to create a package from a proper folder structure (an `__init__.py` and minimal `setup.py` file). Doing this will make your imports cleaner in the notebooks anyways and you don't need to redo anything after switching to create a package. What do others think?

**Comment by k8hertweck** [comment]
1. The project structure chapter does briefly mention including a README in all directories. I like assigning such things as a challenge, which allows people to put as much or little work into it as they would like

**Comment by k8hertweck** [comment]
3. I'm also waffling about how the activities in the project structure (and then packaging) chapter will work. It makes sense to me to set things up to require minimal changes later. If that's the case, the actual changes to the project directory during the project structure chapter may be restricted to adding the READMEs, changelog, etc.

**Comment by joelostblom** [comment]
> 1. I'm also waffling about how the activities in the project structure (and then packaging) chapter will work. It makes sense to me to set things up to require minimal changes later. If that's the case, the actual changes to the project directory during the project structure chapter may be restricted to adding the READMEs, changelog, etc. Additionally, this could involve creating folders that are common (but not specifically used) in this project, like notebooks/

Yes, I think that makes sense also. We could put things in the proper directories from the get go and then add what is missing and explain more in depth during the project chapter. Alternatively, we could move the project structure chapter to be in the beginning. I am good with either of these approaches.

**Comment by k8hertweck** [comment]
I also toyed with the idea of moving the project structure chapter earlier in the book. I'm inclined to leave it in its current location because:
- setting up the project structure properly from the start normalizes the approach, and also makes it easier to explain alternatives (for other projects or approaches)
- dealing with a bunch of empty folders and stubs for files in the git chapters would be tricky

I'm happy to make another pass through the earlier parts of the book and adjust things as needed.

**Comment by lwjohnst86** [comment]
This discussion is exactly why I changed things around in the R RSE narrative (PR #337) and from the beginning am emphasizing creating an R package as the main learning outcome. So that we'd start creating the R package setup at the beginning and not later. The nice thing is that the R package file/folder setup works really nicely with non-package projects too. So I am 100% in agreement with item 3.

Anyway, sadly @joelostblom 

> "Ideally we would be able to suggested an almost language independent dir structure"

that is not the case, and I don't think it ever will be. Too many different opinions and ways that languages work for that to happen. The folder structure for R would look different from what you've listed:

```
pkgname/
├── data/
├── data-raw/
├── inst/  (optional folder used if you want specific files accessible within the *inst*alled package
├── man/ (automatically generated folder and files based on function documentation)
├── R/
├── tests/
├── vignettes/  (This contains the "reports"/R Markdown files)
├── LICENSE.md
├── README.md
├── DESCRIPTION  (Contains the metadata and package dependencies)
├── ... plus several other files related to R package development
└── .Rproj
```

I much more like this folder layout since its pretty clean and simple. 

- All docs are in `vignettes/` (could be `doc/` for non-package projects, since having R Markdown files in `vignettes/` has some ... quirks when used in non-package project settings). 
- All R functions in `R/`.
- Processed data in `data/`, raw data and processing R scripts in `data-raw/`.

The only folders and files that are completely required to make use of R package development workflows (that are applicable to non-package projects too) are the `R/` folder and the `DESCRIPTION` file.

**Comment by joelostblom** [comment]
@lwjohnst86 I think this looks great! There are mostly minor differences, some of which we might be able to erase:

1. Would you also include `CHANGELOG.md` at top level?
2. Is there an advantage of having `data-raw` at top level instead of having it as the subdirectory `data/raw`?
3. Where does user facing documentation go in R, is this is `man` or in `vignettes`? My limited experience has been that docs and tutorials are often written as PDFs and put in vignettes, but there might be other ways?
4. Related to the above, is `vignettes` for output files and their source files? Are the source files Rmarkdown or just R-scripts?
5. Where would you put notebooks with exploratory data analysis? Are these also in vignettes mixed with user facing documentation? Are these ever put in a top level directory such as `notebooks` (and then their output, such as HTML goes in vignettes?)?

The summary of all the above is: could an R project structure look like the
below?

```
pkg_name/
├── data/
│   ├── processed/
│   └── raw/
├── inst/
├── man/
├── notebooks/
├── R/
├── tests/
├── vignettes/
│   └── fig/
├── CHANGELOG.md
├── DESCRIPTION
├── LICENSE.txt
├── README.md
├── NAMESPACE
└── .Rproj
```

With the Python equivalents in parenthesis where the naming is different, but
the content is similar (I removed `setup.cfg` since this is not needed):

```
pkg_name/
├── data/
│   ├── processed/
│   └── raw/
├── inst/          # no equivalent
├── man/           (doc)
├── notebooks/
├── R/             (pkg_name or src/pkg_name)
├── tests/
├── vignettes/     (reports)
│   └── fig/
├── CHANGELOG.md
├── DESCRIPTION    (setup.py)
├── LICENSE.txt
├── README.md
├── NAMESPACE      # no equivalent
└── .Rproj         (environment.yaml)
```

With the caveat that I am still trying to figure out how similar `vignettes` and `reports` would be as per my questions above (and also `man` and `doc` I guess).

Two more disucssion items for Python:
1. I think we would put all user documentation in doc, whether it is generated from a notebook or from .md/.rst files. Does anyone have recommendation for file structure within `doc`? I often see a folder for tutorials or example notebooks, which I think is a good idea in addition to the more structured user documentation. I can also see an argument for making the entire documentation as sample notebooks, I think the Pyviz projects do this.
2. Happy to discuss another name for "reports" in Python is something else is commonly used.

**Comment by joelostblom** [comment]
@k8hertweck 

> I also toyed with the idea of moving the project structure chapter earlier in the book. I'm inclined to leave it in its current location because:
>
>    - setting up the project structure properly from the start normalizes the approach, and also makes it easier to explain alternatives (for other projects or approaches)
>    - dealing with a bunch of empty folders and stubs for files in the git chapters would be tricky

I agree with this, give them only the necessary directories for each chapter but in the right structure so that we don't have to move later, mention that we will make a package and then provide additional dirs and explanations in the proj_struct chapter.

> I'm happy to make another pass through the earlier parts of the book and adjust things as needed.

Thank you!

**Comment by lwjohnst86** [comment]
1. Yes, but it's called `NEWS.md` instead.
2. Its due to the fact that `data/` is built with the package, so any data files will be included in the installed package. But `data-raw/` won't be installed with the package. So it needs to be separate. It's also part of the devtools/usethis workflow, so best go with established norms here.
3. Yes, user facing docs are put in `vignettes/` and `man/`. The PDF/HTML versions of vignettes are built in the same location as the R Markdown document (aka the `vignettes/`). There is some nuance here related to being optimal in package build speed and size when it comes to the vignettes, and hence the reason I said for non-package projects using `doc/` is fine. The documentation in `man/` is actually built from the roxygen documents linked to the function itself (kinda like the python docstrings). You don't actually edit any of the files in `man/`.
4.  Yes, as I mentioned above, `vignettes/` contain the Rmd, html/pdf, and R files associated with the documentation files. Output and source are all together.
5. Notebooks aren't really that common, in my experience. Or at least they are indistinguishable from the Rmd file. So you won't have a separate folder for them. For exploratory stuff, yea, maybe you'd use `use_article()` which creates a "non-vignette" file inside of `vignettes/`. But honestly, for RSE work, you wouldn't really have exploratory docs, or at least you won't commit them to Git. For my non-package projects, I put manuscript and exploratory docs in the same folder in `doc/`.

So it would be:

```
pkgname/
├── data/
├── data-raw/
├── inst/
├── man/
├── R/
├── tests/
│   └── testthat/ (for using the testthat package testing suite)
├── vignettes/
│   └── fig/ (I guess you could have this, but depends on how the figures are used)
├── NEWS.md
├── DESCRIPTION
├── LICENSE.md
├── README.md
├── NAMESPACE
└── pkgname.Rproj
```

Comments:

- Notice the removed `_` in package name. R packages can't have underscores, spaces (obvs), or minus (like python). It's recommended not to have dots either.
- Figures generally are created from functions when in Rmd files. But for non-function generated figures or those that take a long time to render, they could be in `vignettes/fig/` or in `man/figures/`. Either of these work and will be included when building the package.

**Comment by joelostblom** [comment]
Thanks for the detailed reply! This is great for me to understand the
differences between the languages and being able to explain the project
structures well.

1. :+1:
2. Ah ok, would you also create `data-processed` then? When a time consuming operation is applied on the raw data to create a processed/interim data set that is used for downstream applications (possibly multiple scripts/notebooks)? I realize the answer to this might relate to point 5 below.
3. Got it. It sounds like `vignette` is more similar to `doc` in Python then (source and output of documentation files). Then `man` (autogenerated docs) does not have a Python equivalent.
4. :+1:
5. This is interesting. I am approaching RSE as the same file structure will be used as for exploratory work. In brief, the general workflow I use for EDA is to start out in a notebook and as the code grows beyond simple operations, I create functions. When I have more than a few short functions, I extract these from the notebook and put them in .py files inside the `pkg_name` subdir (rather than creating a dir for "throwaway" scripts).
    - In python this works well because I can install package as editable, so
      any changes made to the functions in the package is reflected immediately
      in the notebook (this is also what I opened issue #241 to discuss).
    - So, just creating a package would have a subset of the dirs (no data,
      notebooks or reports), and just doing basic EDA without any user made
      functions would have have another subset (no pkg_name, tests, or
      setup.py). Creating an elaborate data analysis project would have all dirs, since the code not in notebooks would go into a package dir.
6. You mentioned using `doc` for notebooks and their output in projects that require data analysis. What do you do when you have a project with both significant functions/scripts and notebooks? Do you break out the package as a separate project and give it a different name? Do you source scripts directly without bothering creating a proper package dir structure? Or do you just add the `doc` dir to the structure we have been discussing here?

---

For only creating a package, the file structure comparison would look like this:

```shell
# py                   # R
pkg_name/              pkgname/
│                      ├── inst/
│                      ├── man/
├── pkg_name/          ├── R/
├── tests/             ├── tests/
│                      │   └── testthat/
├── doc/               ├── vignettes/
├── CHANGELOG.md       ├── NEWS.md
├── setup.py           ├── DESCRIPTION
├── LICENSE.md         ├── LICENSE.md
├── README.md          ├── README.md
│                      ├── NAMESPACE
└── environment.yaml   └── pkgname.Rproj
```

**Comment by cwickham** [comment]
This is great discussion, it's helping me understand the Python side!

Just one comment on `doc/`: an R package might have this folder to hold files for generating a webpage, e.g using [pkgdown](https://pkgdown.r-lib.org/articles/pkgdown.html).  Although, then I'd generally follow the GitHub pages convention of calling it `docs`.  Using `docs` is what we recommend in the [Novice R publishing](https://merely-useful.github.io/r/r-publishing.html) section.

**Comment by lwjohnst86** [comment]
@cwickham, right, but that would be `docs/` with an `s` :stuck_out_tongue: I know, I really don't like GitHub's "docs" folder decision... that you can't even change! In GitLab the default is `public/` (which makes sense, it is public facing!) but you can change the name if you want... But not GitHub :roll_eyes:

**Comment by lwjohnst86** [comment]
@joelostblom 

For 2. No you wouldn't make a `data-processed/` folder. Any additional processing can also be done in `data-raw/` with R scripts in there (or other scripts, doesn't have to be R). For instance, a project I'm currently working on I store the results output from the "processed data" as a dataframe via a function I created in `R/` and save it as a new dataset in `data/`. Appropriately named of course :relaxed: 

For 5 and also 6. You can totally approach non-package projects the same way as you would for package projects. In fact, they're almost identical in the overall workflow, except when it comes to `doc/` vs `vignettes/`. So I'd start with creating an "EDA" type Rmd file in `doc/`, building the code up and as often as possible start converting those code into functions and moving them into `R/`. No "scripts" in the python sense are used outside of `data-raw/`. If I want to run something outside of the Rmd file, I'd put it in the drake build "plan" (e.g. generate an image or results dataset), and reference the built objects directly in the Rmd file. Very quickly notebooks become irrelevant since the functions describe the process better than a notebook might and very quickly the results output can start being added to the manuscript as individual functions (e.g. `create_cluster_analysis_plot()`).

The general workflow for this all is the same as if I was making an R package... adding package dependencies to DESCRIPTION, writing roxygen function documentation, "exporting, importing" function from other packages, "loading/attaching" the project as if it was an installed package (via `devtools::load_all()`), and so on. I might not ever get out of this workflow if the project doesn't move into "package-worthy" territory. But if it does, I'd start creating vignettes describing the functionality and basic usage, and I'd slowly start removing the files in `doc/` (since I don't believe a polished package should have any left over EDA type files). No need to break out as a new package folder.

**Comment by joelostblom** [comment]
> For 2. No you wouldn't make a data-processed/ folder. Any additional processing can also be done in data-raw/ with R scripts in there (or other scripts, doesn't have to be R). For instance, a project I'm currently working on I store the results output from the "processed data" as a dataframe via a function I created in R/ and save it as a new dataset in data/. Appropriately named of course relaxed

I think I understand, just to confirm: that would mean that all processed data is included in the package when building it? But the increased size from this would only be an issue for public packages, not local ones.

> For 5 and also 6. You can totally approach non-package projects the same way as you would for package projects. In fact, they're almost identical in the overall workflow, except when it comes to doc/ vs vignettes/. So I'd start with creating an "EDA" type Rmd file in doc/, building the code up and as often as possible start converting those code into functions and moving them into R/. 

Great! This sounds very similar to the workflow I described, which I hope means that we can explain the overall concepts the same between languages and make life easier both for learners and ourselves.

> Very quickly notebooks become irrelevant since the functions describe the process better than a notebook might and very quickly the results output can start being added to the manuscript as individual functions (e.g. create_cluster_analysis_plot()).

Agree that notebooks rapidly become a less ideal place to store code for functions. I slightly differ on their role vs the role of a manuscript. I would leave exploratory notebooks as is, even after creating a manuscript. They document all the exploratory analysis I have done, while the most important insights from the notebooks would go into a final product such as a manuscript. There are also many projects that don't culminate in an official manuscript, but the purpose is incremental communication and documentation throughout, where notebooks excel in my opinion.

> I might not ever get out of this workflow if the project doesn't move into "package-worthy" territory. But if it does, I'd start creating vignettes describing the functionality and basic usage, and I'd slowly start removing the files in doc/ (since I don't believe a polished package should have any left over EDA type files). No need to break out as a new package folder.

I agree that EDA-like notebooks should not be part of published packages, there should only be tutorial-like notebooks that are useful for the users of that package. I do think EDA/analysis notebooks are still important for local packages, which I think is the most commonly occurring scenario. Most of the code we write is specific to projects we use internally/locally, these will have analysis notebooks and their function code will be formalized into local packages rather than unorganized scripts (in line with what we said above regarding the workflow always being to put function code into a package structure). Visualization and documentation of the analysis are still in the notebook.

**Comment by lwjohnst86** [comment]
@joelostblom for the first comment, you got it :) Dataset size only becomes an issue if you want to submit to CRAN (they have a strict size limit of something like 5 Mb or less).

**Comment by joelostblom** [comment]
Luke and I spoke on the phone after our meeting, trying to align the last pieces of the project structure. To us, these are the folders that would be added for an EDA project:

```shell
# py                   # R
pkg_name/              pkgname/
├── data/              │
│   ├── processed/     ├── data/
│   └── raw/           ├── data-raw/
├── reports/           ├── doc/
└── public/            └── public/
```

-   The layout of the `data/` folders has already been discussed earlier in this issue.
-   `public/` would be for content that is to be shared online, e.g. a website or HTML versions on notebooks that are shared via an online link. This folder could be called `docs/` as per what @cwickham wrote above regarding GithHub, I am calling it public here to emphasize its intention.
-   `reports/` (for py) and `doc/` (for R) is where I imagine most discussion could be had.
    -   Luke usually stores Rmarkdown notebooks in this folder as well as manuscript and posters (both their source and rendered output if there is any). Figures are usually not stored as images anywhere, but created from a function which is called in an .Rmd file (if they are stored, would be in `doc/fig/`.
    -   My structure is similar, I store my notebooks and their rendered HTML versions in a folder named `notebooks/`. I called it `reports/` here because it is more general and could include md/Rmd docs as well for longer manuscripts or posters (maybe not that usual outside science, probably most reporting is in Jupyter notebooks for Python). For big projects with many notebooks or manuscript items, subfolders could be used, for small projects this is probably not necessary, what do people think? Happy to discuss other names of this folder if there are suggestions. Note that the Python data science cookiecutter (linked above) uses `notebooks/` (and then `reports/` for their output), so let me know if you rather prefer to that approach. Figures can be stored under a `fig/` subfolder or never stored and always created from a function.
-   Note that the above is for RMarkdown and Jupyter notebooks used for analysis. Any notebooks used for user facing documentation and demonstration would go in the documentation folder (`vignettes/` for R and `doc/` for Python).
-   This means that `doc/` means "documents" in R projects and "documentation" in Python projects. A minor discrepancy.

What do you all think about this suggested structure and the comments above? Particularly everyone who were part of the discussion in the meeting yesterday @cwickham @k8hertweck  @brandeism  @DamienIrving.

**Comment by joelostblom** [comment]
I didn't see any further discussion about this in the notes from last week's meeting. Are there any additional comments on this before I implement it in the PRs I have open now? Maybe @gvwilson also in case you have points you think we have missed.

The aspect I think the most about is what is the best solution to the notebooks/reports folder structure/substructure and naming. I still think what is suggested in the previous comment makes the most sense, but happy to discuss more if there are ideas.

**Comment by joelostblom** [comment]
I forgot to say this on the meeting today, but if we want to start with the project structure suggested here, we need an issue to rename the `bin` folder in earlier chapters.

**Comment by k8hertweck** [comment]
@joelostblom I'm working on that, will be included in #363
